instance/models/instance.py

```diff
@@ -0,0 +1,436 @@
+
+sha1_validator = RegexValidator(regex='^[0-9a-f]{40}$', message='Full SHA1 hash required')
+
+
+# Models ######################################################################
+
+class Instance(ValidateModelMixin, TimeStampedModel):
+    """
+    Instance - Group of servers running an application made of multiple services
+    """
+    sub_domain = models.CharField(max_length=50)
+    email = models.EmailField(default='contact@example.com')
+    name = models.CharField(max_length=250)
+
+    base_domain = models.CharField(max_length=50, default=settings.INSTANCES_BASE_DOMAIN)
+    protocol = models.CharField(max_length=5, default='http', choices=PROTOCOL_CHOICES)
+
+    last_provisioning_started = models.DateTimeField(blank=True, null=True)
+
+    class Meta:
+        abstract = True
+        unique_together = ('base_domain', 'sub_domain')
+
+    def __str__(self):
+        return '{0.name} ({0.url})'.format(self)
+
+    @property
+    def domain(self):
+        """
+        Instance domain name
+        """
+        return '{0.sub_domain}.{0.base_domain}'.format(self)
+
+    @property
+    def url(self):
+        """
+        Instance URL
+        """
+        return u'{0.protocol}://{0.domain}/'.format(self)
+
+    @property
+    def active_server_set(self):
+        """
+        Returns the subset of `self.server_set` which aren't terminated
+        """
+        return self.server_set.exclude_terminated()
+
+    @property
+    def status(self):
+        """
+        Instance status
+        """
+        active_server_set = self.active_server_set
+        if not active_server_set:
+            return 'empty'
+        elif active_server_set.count() > 1:
+            raise InconsistentInstanceState('Multiple servers are active, which is unsupported')
+        else:
+            return active_server_set[0].status
+
+
+# Git #########################################################################
+
+class VersionControlInstanceMixin(models.Model):
+    """
+    Instances linked to a VCS, such as git
+    """
+
+    class Meta:
+        abstract = True
+
+    branch_name = models.CharField(max_length=50, default='master')
+    ref_type = models.CharField(max_length=50, default='heads')
+    commit_id = models.CharField(max_length=40, validators=[sha1_validator])
+
+    @property
+    def commit_short_id(self):
+        """
+        Short `commit_id`, limited to 7 characters like on GitHub
+        """
+        if not self.commit_id:
+            return None
+        return self.commit_id[:7]
+
+
+class GitHubInstanceMixin(VersionControlInstanceMixin):
+    """
+    Instance linked to a GitHub repository
+    """
+    github_organization_name = models.CharField(max_length=50, db_index=True)
+    github_repository_name = models.CharField(max_length=50, db_index=True)
+    github_admin_organization_name = models.CharField(max_length=50, blank=True,
+                                                      default=settings.DEFAULT_ADMIN_ORGANIZATION)
```

Comment:
@antoviaque: from my experience, 50 chars is not enough for such kinds of data :) 200 is more or less enough. Also, might make sense to use constants for common field lengths.